Common/include/geometry_structure.hpp

```diff
@@ -1535,6 +1535,44 @@ class CGeometry {
    */
   void SetCustomBoundaryHeatFlux(unsigned short val_marker, unsigned long val_vertex, su2double val_customBoundaryHeatFlux);
   
+  /*!
+   * \brief Filter values given at the element CG by performing a weighted average over a radial neighbourhood.
+   * \param[in] filter_radius - Parameter defining the size of the neighbourhood.
+   * \param[in] kernels - Kernel types and respective parameter, size of vector defines number of filter recursions.
+   * \param[in] input_values - "Raw" values.
+   * \param[out] output_values - Filtered values.
+   */
+  void FilterValuesAtElementCG(const vector<su2double> filter_radius, const vector<pair<unsigned short,su2double> > &kernels,
+                               const su2double *input_values, su2double *output_values) const;
+  
+  /*!
+   * \brief Build the global (entire mesh!) adjacency matrix for the elements in compressed format.
+   *        Used by FilterValuesAtElementCG to search for geometrically close neighbours.
+   * \param[out] neighbour_start - i'th position stores the start position in "neighbour_idx" for the immediate
+   *             neighbours of global element "i". Size nElemDomain+1
+   * \param[out] neighbour_idx - Global index of the neighbours, mush be NULL on entry and free'd by calling function.
+   */
+  void GetGlobalElementAdjacencyMatrix(vector<unsigned long> &neighbour_start, long *&neighbour_idx) const;
```

Comment:
We have to build the adjacency info for all points (not elems) during preprocessing that we then send to ParMETIS for graph partitioning. This is done in parallel and then the missing info is communicated around to the other procs. I know it is different, but perhaps the code there can give you some ideas for how you can adapt for element adjacency and parallelize/communicate. Just in case it is useful, see the routines related to Partition/Distribute/Load in geometry_structure.cpp, starting with CPhysicalGeometry::DistributeColoring(). If the filters don't change during runtime, at least you will only need to do this once at the beginning.